contracts/ProtocolReserve/ProtocolShareReserve.sol

```diff
@@ -33,56 +94,172 @@ contract ProtocolShareReserve is Ownable2StepUpgradeable, ExponentialNoError, Re
+                if (config.destination == _config.destination) {
+                    _config.percentage = config.percentage;
+                    updated = true;
+                }
+                total += _config.percentage;
+            }
+        }
+
+        if (!updated) {
+            total += config.percentage;
+            distributionTargets.push(config);
+        }
+
+        require(total <= MAX_PERCENT, "ProtocolShareReserve: Percentage must be between 0 and 100");
+    }
+
+    /**
+     * @dev Used to find out the amount of funds that's going to be released when release funds is called.
+     * @param comptroller the comptroller address of the pool
+     * @param schema the schema of the distribution target
+     * @param destination the destination address of the distribution target
+     * @param asset the asset address which will be released
+     */
+    function getUnreleasedFunds(
+        address comptroller,
+        Schema schema,
+        address destination,
+        address asset
+    ) external view returns (uint256) {
+        for (uint i = 0; i < distributionTargets.length; i++) {
+            DistributionConfig storage _config = distributionTargets[i];
+            if (_config.schema == schema && _config.destination == destination) {
+                uint256 total = assetsReserves[comptroller][asset][schema];
+                return (total * _config.percentage) / MAX_PERCENT;
+            }
+        }
     }
 
     /**
      * @dev Release funds
-     * @param asset  Asset to be released
-     * @param amount Amount to release
-     * @return Number of total released tokens
+     * @param assets assets to be released to distribution targets
      */
-    function releaseFunds(address comptroller, address asset, uint256 amount) external returns (uint256) {
-        require(asset != address(0), "ProtocolShareReserve: Asset address invalid");
-        require(amount <= poolsAssetsReserves[comptroller][asset], "ProtocolShareReserve: Insufficient pool balance");
+    function releaseFunds(address comptroller, address[] memory assets) external {
+        _accruePrimeInterest();
+        
+        for (uint i = 0; i < assets.length; i++) {
+            _releaseFund(comptroller, assets[i]);
+        }
+    }
+
+    function _releaseFund(address comptroller, address asset) internal {
+        uint256 schemaOneBalance = assetsReserves[comptroller][asset][Schema.ONE];
+        uint256 schemaTwoBalance = assetsReserves[comptroller][asset][Schema.TWO];
+        uint256 schemaOneTotalTransferAmount;
+        uint256 schemaTwoTotalTransferAmount;
 
-        assetsReserves[asset] -= amount;
-        poolsAssetsReserves[comptroller][asset] -= amount;
-        uint256 protocolIncomeAmount = mul_(
-            Exp({ mantissa: amount }),
-            div_(Exp({ mantissa: protocolSharePercentage * expScale }), baseUnit)
-        ).mantissa;
+        for (uint i = 0; i < distributionTargets.length; i++) {
```

Comment:
We can use a temporary variable to store distributionTargets.length, to save gas. Moreover, we can use ++i instead of i++, also to save some gas